browser/tor/onion_location_navigation_throttle.cc

```diff
@@ -0,0 +1,112 @@
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/tor/onion_location_navigation_throttle.h"
+
+#include <string>
+#include <utility>
+
+#include "base/bind.h"
+#include "brave/browser/profiles/profile_util.h"
+#include "brave/browser/tor/onion_location_tab_helper.h"
+#include "brave/browser/tor/tor_profile_service.h"
+#include "brave/common/tor/pref_names.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/profiles/profile_window.h"
+#include "chrome/browser/ui/browser.h"
+#include "chrome/browser/ui/browser_finder.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/web_contents.h"
+
+namespace tor {
+
+namespace {
+
+bool GetOnionLocation(const net::HttpResponseHeaders* headers,
+                      std::string* onion_location) {
+  onion_location->clear();
+  std::string name = "onion-location";
+
+  if (!headers || !headers->EnumerateHeader(nullptr, name, onion_location))
+    return false;
+  return true;
+}
+
+void OnTorProfileCreated(GURL onion_location,
+                         Profile* profile,
+                         Profile::CreateStatus status) {
+  if (status != Profile::CreateStatus::CREATE_STATUS_INITIALIZED)
+    return;
+  Browser* browser = chrome::FindTabbedBrowser(profile, true);
+  if (!browser)
+    return;
+  content::OpenURLParams open_tor(onion_location, content::Referrer(),
+                                  WindowOpenDisposition::OFF_THE_RECORD,
+                                  ui::PAGE_TRANSITION_TYPED, false);
+  browser->OpenURL(open_tor);
+}
+
+}  // namespace
+
+// static
+std::unique_ptr<OnionLocationNavigationThrottle>
+OnionLocationNavigationThrottle::MaybeCreateThrottleFor(
+    content::NavigationHandle* navigation_handle) {
+  if (tor::TorProfileService::IsTorDisabled())
+    return nullptr;
+  return std::make_unique<OnionLocationNavigationThrottle>(navigation_handle);
+}
+
+OnionLocationNavigationThrottle::OnionLocationNavigationThrottle(
+    content::NavigationHandle* navigation_handle)
+    : content::NavigationThrottle(navigation_handle) {
+  profile_ = Profile::FromBrowserContext(
+      navigation_handle->GetWebContents()->GetBrowserContext());
+}
+
+OnionLocationNavigationThrottle::~OnionLocationNavigationThrottle() {}
+
+content::NavigationThrottle::ThrottleCheckResult
+OnionLocationNavigationThrottle::WillProcessResponse() {
+  if (navigation_handle()->IsInMainFrame()) {
+    auto* headers = navigation_handle()->GetResponseHeaders();
+    std::string onion_location;
+    if (headers && GetOnionLocation(headers, &onion_location) &&
+        !navigation_handle()->GetURL().DomainIs("onion")) {
+      // If user prefers opening it automatically
+      if (profile_->GetPrefs()->GetBoolean(prefs::kAutoOnionLocation)) {
+        profiles::SwitchToTorProfile(
+            base::BindRepeating(&OnTorProfileCreated, GURL(onion_location)));
+      } else {
+        OnionLocationTabHelper::SetOnionLocation(
```

Comment:
AFAIU there is no way to disable the omnibox view, even if I don't want to know about onions - maybe we should consider this.

Reply:
This will hide `OnionLocationView` when we navigate away to other website without onion-location header.
This feature will be disabled when user disable tor. 
<img width="708" alt="Screen Shot 2020-10-08 at 11 50 58" src="https://user-images.githubusercontent.com/11330831/95501014-8cdf0780-095c-11eb-9da9-848fbbc4f0fe.png">

Or do you mean give users another option that they want enabling tor but don't want to see this "Open in Tor"/"Onion Available" label? If so, that sounds very weird, even in Tor browser bundle they don't provide this kind of option. cc @diracdeltas

Reply:
Yeah, I just want to understand product/UX consequences. There are two facts: (1) "Private window with Tor" pref is enabled by default (2) according to P3A, only 20% of users ever used Tor. 

So by always showing this view (for certain websites) we probably will confuse 80% of users who don't care about onion domains at all, and most likely don't want to have any new buttons in the omnibox. 

On the other hand, only few sites have the onion counterpart, so for now we are probably ok as is.